Refactor organizations

sampaccoud opened the pull request:
## Purpose

This is counter proposal for refactoring organizations and try to make it simpler and more powerful

## Proposal

Organizations were refactored with the following in mind:
- It's easier to start from scratch,
- Try not to rely on PageExtensions but have our own separate Organization model. The trick was to introduce a OneToOne relation from the Organization to a normal Page so we can still make `my_page.organization.logo.url` while relying on pure DjangoCMS pages,
- Add a wizard specific to adding a new Organization page, which creates an Organization model and links it to the newly created page,

Note: ended-up refactoring also the `create_cms_data` command because its design was making it hard to understand, impossible to `really test` and very slow to `test a little bit`,

jmaupetit commented on apps/organization/tests/test_models.py:
```diff
@@ -0,0 +1,81 @@
+        The `page` and `logo` are not required on the model
+        """
+        page = OrganizationFactory(page=None, logo=None)
+        self.assertTrue(page.pk is not None)
```
`self.assertIsNotNone(page.pk)`?

jmaupetit commented on apps/organization/tests/test_models.py:
```diff
@@ -0,0 +1,81 @@
+        """
+        page = OrganizationFactory(page=None, logo=None)
+        self.assertTrue(page.pk is not None)
+        self.assertTrue(page.page is None)
```
`self.assertIsNone(page.page)`?

jmaupetit commented on apps/organization/tests/test_models.py:
```diff
@@ -0,0 +1,81 @@
+        page = OrganizationFactory(page=None, logo=None)
+        self.assertTrue(page.pk is not None)
+        self.assertTrue(page.page is None)
+        self.assertFalse(bool(page.logo))
```
`self.assertIsNone(page.logo)`?

sampaccoud replied:
This one no. An imageField is not None and must be checked as a bool if I understood well.

jmaupetit commented on apps/organization/tests/test_models.py:
```diff
@@ -0,0 +1,81 @@
+
+    def test_organization_str_with_page(self):
+        """
+        The str representation should include mention of the page if it exist
```
`s/exist/exists/`

jmaupetit commented on apps/organization/tests/test_wizards.py:
```diff
@@ -0,0 +1,64 @@
+
+    def test_organization_create_wizards_list(self):
+        """
+        The wizard to create a new Organization page should be present of the wizards list page
```
`s/of/on/`?

jmaupetit commented on apps/organization/tests/test_wizards.py:
```diff
@@ -0,0 +1,64 @@
+
+    def test_organization_wizard(self):
+        """
+        A user we sufficient access rights should be able to create an organization via the
```
`s/we/with/`

sampaccoud replied:
Sorry #dyslexiaagain!

jmaupetit commented on apps/organization/tests/test_wizards.py:
```diff
@@ -0,0 +1,64 @@
+            'page': page,
+            'logo': None
+        }
+        self.assertTrue(page.organization, expected_values)
```
`assertTrue` and not `assertEqual`?

sampaccoud replied:
Wow why is this test working btw

jmaupetit replied:
:man_shrugging: duno

jmaupetit commented on apps/core/helpers.py:
```diff
@@ -0,0 +1,50 @@
+
+def create_i18n_page(content, is_homepage=False, **kwargs):
+    """
+    Creating an multilingual page is not straightforward so we thought we should have a helper
```
`s/an/a/`

jmaupetit replied:
Can you document the expected `content` argument?

jmaupetit commented on apps/core/helpers.py:
```diff
@@ -0,0 +1,50 @@
+    for language in content.keys():
+        add_plugin(
+            placeholder=placeholder,
+            plugin_type='TextPlugin', language=language,
```
The `language` parameter should be set in a new line :wink:

sampaccoud replied:
... and sorted!

jmaupetit commented on apps/core/management/commands/create_cms_data.py:
```diff
@@ -1,143 +1,110 @@
 def create_cms_data():
-    """Create base CMS data"""
+    """
+    Automatically create a base site for working on the project
```
`s/for working/to work/`?

jmaupetit commented on richie/settings.py:
```diff
@@ -182,9 +182,11 @@ class Base(DRFMixin, ElasticSearchMixin, Configuration):
-        ('en', _('en')),
-        ('fr', _('fr')),
+        ('en', _('English')),
+        ('fr', _('French')),
```
:medal_sports:

jmaupetit commented on apps/core/tests/test_pages.py:
```diff
@@ -0,0 +1,36 @@
+        self.assertIn(content['fr'], response.rendered_content)
+
+        # Get the root page in english...
+        root = Page.objects.get_home()
```
You can safely remove this line (`root` has already been initialized with the same value).

jmaupetit commented on Makefile:
```diff
@@ -97,5 +97,5 @@ watch-ts: ## watch changes in TypeScript files
 
 help:
-	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+	@grep -E '^[a-zA-Z0-9_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
```
:+1:

jmaupetit commented on setup.cfg:
```diff
@@ -10,6 +10,9 @@ exclude =
 addopts = -v --cov-report term-missing
+python_files =
+    test_*.py
+    tests.py
```
:+1:

sampaccoud commented on apps/organization/tests/test_admin.py:
```diff
@@ -0,0 +1,88 @@
+        self.assertIn(
+            '<td class="field-view_in_cms"><a href="/en/my-title/" target="_blank">View</a></td>',
+            content,
+        )
```
Because I didn't know it exists lol. Done.

sampaccoud commented on apps/organization/tests/test_admin.py:
```diff
@@ -0,0 +1,88 @@
+        content = response.content.decode()
+        # Check that the page includes all our fields
+        for field in ['page', 'name', 'code', 'logo']:
+            self.assertIn(field, content)
```
hardcoding the fields is superior for a test because it ensures that we don't add or remove fields unintentionally?

jmaupetit replied:
I was pretty sure that you will answer this. This is a decent strategy. But still, the `assertIn` of a generic word as page, name or logo is too weak don't you think? This is why I proposed to assert at least `id="id_foo"`. WDYT?

sampaccoud replied:
Yes I took into account that change!

sampaccoud commented:
@jmaupetit I added some line breaks in tests and fixed all the issues you raised. Feel free to look at it again and let me know if something is still not exactly the way you would like to see it?

jmaupetit commented on apps/organizations/models.py:
```diff
@@ -0,0 +1,91 @@
+    This model should be used to record structured data about the organization whereas the
+    associated page object is where we record the less structured information to display on the
+    page to present the organization.
+    """
```
Why not giving this model a [`UUIDField`](https://docs.djangoproject.com/en/2.0/ref/models/fields/#uuidfield) as primary key? (I'm a big fan of uuids).

sampaccoud replied:
I'm pro uuidfield too.

jmaupetit commented on apps/organizations/models.py:
```diff
@@ -0,0 +1,91 @@
+        TODO: improve uniquess (find related Django blog post)
+        """
+        if self.code:
+            self.code = self.code.upper().strip().replace(' ', '-')
```
Why not using `slugify(self.code).upper() + uuid.uuid4().hex[:6]`?

```python
from uuid import uuid4
from django.utils.text import slugify

# [...]
self.code = '{code}-{id}'.format(
    code=slugify(self.code).upper(),
    id=uuid4().hex[:6]
)
```

https://docs.djangoproject.com/en/2.0/ref/utils/#django.utils.text.slugify

sampaccoud replied:
what's the idea of uuid ? We want to be able to set the code because there are already existing codes. The idea here was just to normalize them a bit.

sampaccoud replied:
yeah using slugify is a great idea.
I had this blogpost in mind when I added the TODO: https://www.b-list.org/weblog/2018/feb/11/usernames/
I was in the plane so did not remember exactly what to do... should I do it now or is it overkill WDYT?

jmaupetit replied:
Oh yes, I forgot this post! Yes, I think it's a bit overkill as we are not dealing with (potentially malicious) usernames, right?

jmaupetit commented on apps/organizations/models.py:
```diff
@@ -0,0 +1,91 @@
+                select_related('extended_object').\
+                get(extended_object__publisher_is_draft=True).\
+                extended_object
+        except OrganizationPage.DoesNotExist:
```
Shouldn't we log something here?

sampaccoud replied:
No it's not abnormal if there is no page. I use the try/except pattern because it is better to ask for forgiveness than permission.

jmaupetit commented on apps/organizations/templates/organizations/cms/organization.html:
```diff
@@ -0,0 +1,11 @@
+    {% endif %}
+{% endwith %}
+{% placeholder "maincontent" %}
+{% endblock content %}
```
We should discuss templates indentation.

sampaccoud replied:
I modified it.

jmaupetit commented on apps/organizations/templates/organizations/cms/organization.html:
```diff
@@ -0,0 +1,11 @@
+<h1>{{ current_page.get_title }}</h1>
+{% with current_page.organizationpage.organization as organization %}
+    {% if organization.logo %}
+    <img src="{{ organization.logo.url }}">
```
Don't forget the `alt` attribute for a minimal accessibility.

jmaupetit commented on apps/organizations/tests/test_admin.py:
```diff
@@ -0,0 +1,102 @@
+    def test_organization_list_view_no_page(self):
+        """
+        When the organization is not related to a CMS page, the amdin list view should display a
+        hyphen instead of the link to the associated page
```
hyphen or "no-page flag"?

sampaccoud replied:
well spotted!

jmaupetit commented on apps/organizations/tests/test_admin.py:
```diff
@@ -0,0 +1,102 @@
+
+        # Check that the page includes all our fields
+        for field in ['name', 'code', 'logo']:
+            self.assertContains(response, 'id_{:s}'.format(field))
```
:medal_sports: :wink:

jmaupetit commented on apps/organizations/tests/test_models_organization.py:
```diff
@@ -0,0 +1,111 @@
+        The `code` field should be normalized to improve its uniqueness
+        """
+        organization = OrganizationFactory(code=' r5G ypm 44 ')
+        self.assertEqual(organization.code, 'R5G-YPM-44')
```
Can we also test with a code containing unicode characters (Emoji, Chinese, ...)?

sampaccoud replied:
Great idea: " r5G yç👷学pm 44 " -> "R5G-YÇ👷学PM-44"

jmaupetit commented on apps/organizations/tests/test_models_organization.py:
```diff
@@ -0,0 +1,111 @@
+        does not exist.
+        """
+        organization = OrganizationFactory(name='Sorbonne', code='SOR')
+        self.assertIsNone(organization.get_page())
```
An empty row is missing at the end of this file.

jmaupetit commented on apps/organizations/tests/test_wizards.py:
```diff
@@ -0,0 +1,140 @@
+        self.assertTrue(form.is_valid())
+        page = form.save()
+        # Check that the slug has been truncated
+        self.assertEqual(page.get_slug(), 't'*200)
```
`self.assertEqual(page.get_slug(), 't' * 200)`?

sampaccoud replied:
done

jmaupetit commented on apps/organizations/tests/test_wizards.py:
```diff
@@ -0,0 +1,140 @@
+        self.assertEqual(form.errors, {'title': [
+            'You must first create an `organization list` page and set its `reverse_id` to '
+            '`organizations`.'
+        ]})
```
Empty row is missing.

sampaccoud replied:
weird because there is an empty row.

jmaupetit replied:
Okay my bad. Forget this comment.

rmoch commented on apps/organizations/admin.py:
```diff
@@ -0,0 +1,35 @@
+        # opens a new tab and displays this page.
+        page = organization.get_page()
+        if page:
+            return '<a href="{url:s}" target="_blank">{anchor!s}</a>'.format(
```
It's a ui decision, I would totally assume
Users of Django admin are us, the dev team
When using keyboard I prefer to close a tab, than hit backspace, because some widget could have take focus, and my right hand is still on the mouse
you ?

jmaupetit replied:
I usually CTR+Click to open the link in a new tab and not having to deal with my current tab history. But it's my choice, my way of organizing the way I browse/work. Nobody forced me to do it that way.

sampaccoud replied:
I have no strong opinion here. 
I always use "target=_blank" for links inside django admin, because it is a tool that people use all day long and they most probably don't want you to mess with this tab. 
My feeling is that most websites that are tools work like this (gmail for example)

jmaupetit replied:
Let's ask our users then!

jmaupetit replied:
And make this issue non-blocking by declaring an issue to push further the discussion?

sampaccoud commented on apps/organizations/cms_wizards.py:
```diff
@@ -0,0 +1,99 @@
+        """
+        if not Page.objects.filter(
+                reverse_id=ORGANIZATIONS_PAGE_REVERSE_ID,
+                publisher_is_draft=True,
```
pylint was complaining. It can not be on the same level as below's line "raise..."

sampaccoud commented on apps/organizations/models.py:
```diff
@@ -1,55 +1,91 @@
+        max_length=100,
+        unique=True,
+        null=True,
+        blank=True,
```
The idea is that we (FUN) should set codes, not the organization people. So the wizard could be filled by the organization themselves and we later come and add a code.

rmoch replied:
🤔

sampaccoud commented on apps/organizations/models.py:
```diff
@@ -1,55 +1,91 @@
+        """
+        if self.code:
+            self.code = self.code.upper().strip().replace(' ', '-')
+        return super().save(*args, **kwargs)
```
but we will have slugs... so can we normalize them a bit ?

sampaccoud commented on apps/organizations/tests/test_models_organization.py:
```diff
@@ -0,0 +1,111 @@
+        """
+        The `name` field should be limited to 255 characters
+        """
+        OrganizationFactory(name='a'*255)
```
I think testing that it works with the max 255 and not with 256.

rmoch replied:
👌

sampaccoud commented on apps/organizations/tests/test_admin.py:
```diff
@@ -0,0 +1,102 @@
+        self.assertContains(
+            response,
+            '<td class="field-view_in_cms"><a href="/en/my-title/" target="_blank">View</a></td>',
+            status_code=200,
```
Ok let me remove it.

sampaccoud commented on apps/organizations/tests/test_admin.py:
```diff
@@ -0,0 +1,102 @@
+
+    def test_organization_list_view_no_page(self):
+        """
+        When the organization is not related to a CMS page, the amdin list view should display a
```
fixed it.

sampaccoud commented on apps/organizations/tests/test_admin.py:
```diff
@@ -0,0 +1,102 @@
+        # Get the admin change view
+        url = reverse('admin:organizations_organization_change', args=[organization.id])
+        response = self.client.get(url, follow=True)
+
```
Must be a copy/paste fail because actually it's not necessary. Removed it.

sampaccoud commented on apps/organizations/models.py:
```diff
@@ -1,55 +1,91 @@
+            code=self.organization.code,
+            name=self.organization.name,
+            model=self._meta.verbose_name.title(),
+        )
```
Don't know where I read it but verbose_names should be lowercase as Django takes care of capitalizing them whenever it is required.

rmoch replied:
👌

jmaupetit commented on apps/organizations/tests/test_models_organization.py:
```diff
@@ -0,0 +1,112 @@
+        Test with a chinese character, an emoji and a french ç...
+        """
+        organization = OrganizationFactory(code=' r5G yç👷学pm 44 ')
+        self.assertEqual(organization.code, 'R5G-YPM-44')
```
I think we must keep them. You can allow unicode with slugify: `slugify(value, allow_unicode=True)`

sampaccoud replied:
I don't think we should because then come all the problems of unicode for unity cf ubernostrum blogpost. What we want for these codes is pure ascii.

jmaupetit commented on apps/organizations/templates/organizations/cms/organization.html:
```diff
@@ -2,14 +2,13 @@
-<h1>{{ current_page.get_title }}</h1>
+  {% with current_page.organizationpage.organization as organization %}
+    {% if organization.logo %}
+      <img src="{{ organization.logo.url }}">
```
If you don't mind, can you add an `alt` attribute here?

sampaccoud replied:
Done
